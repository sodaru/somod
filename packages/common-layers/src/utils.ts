import { existsSync } from "fs";
import { mkdir, readdir, readFile, writeFile } from "fs/promises";
import { join } from "path";
import { chdir } from "process";
import { CommonLayers, layerLibraries } from "./index";
import rimraf from "rimraf";

const deleteDir = (dir: string): Promise<void> => {
  return new Promise((resolve, reject) => {
    rimraf(dir, err => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
};

/**
 * Generates layer structure and package.json(including dependencies) for all configured layers
 * layer structure will be like {base folder}/layers/{layer name}/nodejs/package.json
 */
export const generate = async () => {
  chdir(join(__dirname, ".."));
  await mkdir("./layers", { recursive: true });
  const rootPackageJsonStr = await readFile("./package.json", {
    encoding: "utf8"
  });

  const rootPackageJson = JSON.parse(rootPackageJsonStr);

  for (const layer in layerLibraries) {
    const dependencies = {};
    layerLibraries[layer]["libraries"].forEach((dependency: string) => {
      const dependencyVersion = rootPackageJson.peerDependencies[dependency];
      if (!dependencyVersion) {
        throw new Error(
          `Dependency ${dependency} in layer ${layer} is not found in root package.json's peerDependencies`
        );
      }
      dependencies[dependency] = dependencyVersion;
    });
    const packageJSON = {
      name: `slplayer-${layer.toLowerCase()}`,
      version: "1.0.0",
      description: "Function layer generated by @somod/common-layers",
      dependencies: dependencies
    };
    await mkdir(`./layers/${layer}/nodejs`, { recursive: true });
    await writeFile(
      `./layers/${layer}/nodejs/package.json`,
      JSON.stringify(packageJSON, null, 2)
    );
  }
};

/**
 * compares and removes duplicate libraries from passed layer w.r.t base layer
 * @param layerName layer name for comparing duplicate libraries
 */
export const removeBaseLayerLibraries = async (
  layerName: string
): Promise<void> => {
  chdir(join(__dirname, ".."));
  const baseLayerNodeModulesPath = `./layers/${CommonLayers.baseLayer}/nodejs/node_modules`;
  const layerNodeModulesPath = `./layers/${layerName}/nodejs/node_modules`;
  const dirs = await readdir(layerNodeModulesPath);
  const scopes: string[] = [];
  const libraries = dirs.filter(dir => {
    if (dir.startsWith("@")) {
      scopes.push(dir);
      return false;
    }
    return true;
  });

  await Promise.all(
    scopes.map(async scope => {
      const scopedLibraries = await readdir(`${layerNodeModulesPath}/${scope}`);
      libraries.push(...scopedLibraries.map(l => `${scope}/${l}`));
    })
  );

  await Promise.all(
    libraries.map(async lib => {
      if ([".bin", ".package-lock.json"].includes(lib)) {
        return;
      }
      if (existsSync(`${baseLayerNodeModulesPath}/${lib}`)) {
        const packageJsonStrList = await Promise.all([
          readFile(`${baseLayerNodeModulesPath}/${lib}/package.json`, {
            encoding: "utf8"
          }),
          readFile(`${layerNodeModulesPath}/${lib}/package.json`, {
            encoding: "utf8"
          })
        ]);
        const [baseLayerLibVersion, layerLibVersion] = packageJsonStrList.map(
          p => JSON.parse(p).version
        );
        if (baseLayerLibVersion == layerLibVersion) {
          await deleteDir(`${layerNodeModulesPath}/${lib}`);
        }
      }
    })
  );
};
