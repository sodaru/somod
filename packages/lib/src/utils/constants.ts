export const file_packageJson = "package.json";
export const file_packageLockJson = "package-lock.json";

export const file_dotenv = ".env";

export const file_tsConfigSomodJson = "tsconfig.somod.json";

export const file_nextConfigJs = "next.config.js";

export const key_somod = "somod";

export const key_type = "type";

export const key_module = "module";
export const key_typings = "typings";

export const key_sideEffects = "sideEffects";

export const key_main = "main";
export const key_jsnextMain = "jsnext:main";
export const key_files = "files";

export const key_serverlessSchema = "serverlessSchema";

export const key_somodPlugins = "somodPlugins";

export const path_nodeModules = "node_modules";

export const path_somodWorkingDir = ".somod";

export const path_build = "build";

export const path_ui = "ui";
export const path_lib = "lib";

export const path_pages = "pages";
export const path_pagesData = "pages-data";
export const path_public = "public";

export const file_index_js = "index.js";
export const file_index_dts = "index.d.ts";

export const path_git = ".git";

export const file_configYaml = "config.yaml";
export const file_configJson = "config.json";

export const file_extensionTs = "extension.ts";
export const file_extensionJs = "extension.js";

export const path_serverless = "serverless";
export const path_functions = "functions";
export const path_functionLayers = "functionLayers";
export const path_middlewares = "middlewares";
export const file_templateYaml = "template.yaml";
export const file_templateJson = "template.json";

export const file_samConfig = "samconfig.toml";

export const path_vercel = ".vercel";

export const cli_version_regex =
  /^([0-9]+)\.([0-9]+)\.([0-9]+)(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?(?:\+[0-9A-Za-z-]+)?$/;

export const namespace_page = "UI Page";

export const namespace_public = "UI Public Asset";

export const namespace_env_config = "UI Env Config";

export const namespace_public_runtime_config = "UI Public Runtime Config";

export const namespace_server_runtime_config = "UI Server Runtime Config";

export const namespace_api_gateway = "Serverless Api Gateway";

export const namespace_parameter = "Parameter";

export const namespace_output = "Serverless Output";

export const resourceType_Function = "AWS::Serverless::Function";
export const resourceType_FunctionLayer = "AWS::Serverless::LayerVersion";
export const resourceType_FunctionMiddleware =
  "SOMOD::Serverless::FunctionMiddleware";

export const custom_resource_prefix = "Custom::";

export const file_parametersYaml = "parameters.yaml";

export const file_parametersJson = "parameters.json";

export const key_parameter_overrides = "parameter_overrides";
