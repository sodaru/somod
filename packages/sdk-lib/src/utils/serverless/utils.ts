import { createHash } from "crypto";
import { cloneDeep, isArray, isPlainObject, isUndefined } from "lodash";
import { KeywordAll, SLPKeyword, SLPTemplate } from "./types";

const getSLPKeywordPaths = (chunk: unknown): SLPTemplate["keywordPaths"] => {
  const keywordPaths = Object.fromEntries(
    KeywordAll.map(keyword => [keyword, []])
  ) as SLPTemplate["keywordPaths"];

  if (isPlainObject(chunk)) {
    KeywordAll.forEach(keyword => {
      if (!isUndefined(chunk[keyword])) {
        keywordPaths[keyword].push([]);
      }
    });

    Object.keys(chunk).forEach(key => {
      const childKeywordPaths = getSLPKeywordPaths(chunk[key]);
      KeywordAll.forEach(keyword => {
        childKeywordPaths[keyword].forEach((_keywordPaths: string[]) => {
          _keywordPaths.unshift(key);
          keywordPaths[keyword].push(_keywordPaths);
        });
      });
    });
  } else if (isArray(chunk)) {
    chunk.forEach((arrayItem, index) => {
      const childKeywordPaths = getSLPKeywordPaths(arrayItem);
      KeywordAll.forEach(keyword => {
        childKeywordPaths[keyword].forEach((_keywordPaths: string[]) => {
          _keywordPaths.unshift(index + "");
          keywordPaths[keyword].push(_keywordPaths);
        });
      });
    });
  }
  return keywordPaths;
};

export const updateKeywordPathsInSLPTemplate = (
  slpTemplate: SLPTemplate
): void => {
  const slpKeywords = getSLPKeywordPaths(slpTemplate.original.Resources);
  slpTemplate.keywordPaths = slpKeywords;
};

/**
 * returns the value present in the path (usually Keywords path), and it always looks in original
 * @param slpTemplate : single slp tamplate
 * @param path : json path in array format (usually Keywords path)
 * @returns value found inside the json path
 *          for example : { "SLP::Ref": { module:"@sodaru/slp", "resource": "BaseRestApi" } }
 */
export const getSLPKeyword = <T extends SLPKeyword>(
  slpTemplate: SLPTemplate,
  path: string[]
): T => {
  // always return the keyword value from original
  let keyword: unknown = slpTemplate.original.Resources;

  path.forEach(pathSegment => {
    keyword = keyword[pathSegment];
  });
  return cloneDeep(keyword) as T;
};

export const replaceSLPKeyword = (
  slpTemplate: SLPTemplate,
  path: string[],
  newValue: unknown
): void => {
  let keyword: unknown = slpTemplate.Resources;
  const _path = [...path];
  const lastPathSegment = _path.pop();
  _path.forEach(pathSegment => {
    keyword = keyword[pathSegment];
  });
  keyword[lastPathSegment] = newValue;
};

const hashModuleName = (str: string): string => {
  return createHash("sha256").update(str).digest("hex").substring(0, 8);
};

export const getSAMParameterName = (
  moduleName: string,
  slpParameterName: string
): string => {
  return "p" + hashModuleName(moduleName) + slpParameterName;
};

export const getSAMResourceLogicalId = (
  moduleName: string,
  slpResourceId: string
): string => {
  return "r" + hashModuleName(moduleName) + slpResourceId;
};

export const getSAMResourceName = (
  moduleName: string,
  slpResourceName: string
): unknown => {
  return {
    "Fn::Sub": [
      "slp${stackId}${moduleHash}${slpResourceName}",
      {
        stackId: {
          "Fn::Select": [2, { "Fn::Split": ["/", { Ref: "AWS::StackId" }] }]
        },
        moduleHash: hashModuleName(moduleName),
        slpResourceName
      }
    ]
  };
};
