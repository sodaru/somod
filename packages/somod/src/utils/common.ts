import { Command } from "nodejs-cli-runner";

export type SOMODCommandTypeOptions = {
  ui: boolean;
  serverless: boolean;
};

export type DebugModeOptions = {
  debug: boolean;
};

export const addSOMODCommandTypeOptions = (command: Command) => {
  command.option("--ui", "only ui");
  command.option("--serverless", "only serverless");
};

export const addDebugOptions = (command: Command) => {
  command.option("-d, --debug", "Enable Debug mode");
};

export const getSOMODCommandTypeOptions = (
  options: SOMODCommandTypeOptions
): SOMODCommandTypeOptions => {
  const ui = !options.serverless || options.ui;
  const serverless = !options.ui || options.serverless;
  return { ui, serverless };
};
