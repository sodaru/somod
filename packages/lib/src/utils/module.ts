import { existsSync } from "fs";
import { bfs, Node } from "graph-dsa";
import { uniq } from "lodash";
import { dirname, join } from "path";
import { IModuleHandler, Module, ModuleNode } from "somod-types";
import { path_nodeModules } from "./constants";
import { freeze } from "./freeze";
import { read } from "./packageJson";

export class ModuleHandler implements IModuleHandler {
  private static instance: IModuleHandler;

  private _rootModuleName: string;
  private _moduleNodesMap: Record<string, ModuleNode>;
  private _moduleNamesInBFSOrder: string[];

  private constructor() {
    // do nothing
  }

  public static async getInstance(dir: string): Promise<IModuleHandler> {
    if (this.instance === undefined) {
      const moduleHandler = new ModuleHandler();

      const { rootModuleName, modules, moduleDependencies } =
        await this._loadModules(dir);

      if (rootModuleName === undefined) {
        throw new Error(`${dir} is not a SOMOD module`);
      }

      this._checkForRepeatedModules(modules);

      moduleHandler._rootModuleName = rootModuleName;
      moduleHandler._moduleNodesMap = this._generateModuleNodeMap(
        modules,
        moduleDependencies
      );
      moduleHandler._moduleNamesInBFSOrder = this._sortModules(
        moduleHandler._moduleNodesMap
      );
      freeze(moduleHandler._moduleNodesMap);
      freeze(moduleHandler._moduleNamesInBFSOrder);

      this.instance = moduleHandler;
    }
    return this.instance;
  }

  private static async _loadModules(dir: string) {
    const modules: Module[] = [];
    const moduleDependencies: { moduleName: string; dependencies: string[] }[] =
      [];
    let rootModuleName: string;

    const jobs: Record<string, Promise<void>> = {};
    const pendingJobsQueue: Promise<void>[] = [];

    const loadModule = async (packageDir: string, root = false) => {
      const packageJson = await read(packageDir);
      if (packageJson.somod !== undefined) {
        const module: Module = {
          name: packageJson.name as string,
          version: packageJson.version as string,
          packageLocation: packageDir,
          root
        };
        if (root) {
          rootModuleName = module.name;
        }
        modules.push(module);

        const allDependencies = uniq([
          ...Object.keys(packageJson.dependencies || {}),
          ...(root
            ? Object.keys(packageJson.devDependencies || {})
            : Object.keys(packageJson.peerDependencies || {}))
        ]);

        moduleDependencies.push({
          moduleName: module.name,
          dependencies: allDependencies
        });
        allDependencies.map(dependency => {
          const dependencyPackageLocation = this._findModulePackageLocation(
            dependency,
            packageDir
          );
          if (jobs[dependencyPackageLocation] === undefined) {
            jobs[dependencyPackageLocation] = loadModule(
              dependencyPackageLocation
            );
            pendingJobsQueue.push(jobs[dependencyPackageLocation]);
          }
        });
      }
    };

    jobs[dir] = loadModule(dir, true);
    pendingJobsQueue.push(jobs[dir]);

    while (pendingJobsQueue.length > 0) {
      const awaitingJobQueue = [];
      while (pendingJobsQueue.length > 0) {
        awaitingJobQueue.push(pendingJobsQueue.shift());
      }
      await Promise.all(awaitingJobQueue);
    }

    return { rootModuleName, modules, moduleDependencies };
  }

  private static _findModulePackageLocation(moduleName: string, from: string) {
    let moduleContainingDir = from;
    while (
      !existsSync(join(moduleContainingDir, path_nodeModules, moduleName))
    ) {
      const parentDir = dirname(moduleContainingDir);
      if (parentDir == moduleContainingDir) {
        throw new Error(`Module ${moduleName} not found from ${from}`);
      }
      moduleContainingDir = parentDir;
    }
    return join(moduleContainingDir, path_nodeModules, moduleName);
  }

  private static _checkForRepeatedModules(modules: Module[]) {
    const moduleLocations: Record<string, string[]> = {};
    modules.forEach(module => {
      if (moduleLocations[module.name] === undefined) {
        moduleLocations[module.name] = [];
      }
      moduleLocations[module.name].push(module.packageLocation);
    });

    const repeatedModuleNames = Object.keys(moduleLocations).filter(
      moduleName => moduleLocations[moduleName].length > 1
    );

    if (repeatedModuleNames.length > 0) {
      throw new Error(
        "Following modules are repeated\n" +
          repeatedModuleNames.map(
            repeatedModuleName =>
              `${repeatedModuleName}\n${moduleLocations[repeatedModuleName]
                .map(l => " - " + l)
                .join("\n")}`
          )
      );
    }
  }

  private static _generateModuleNodeMap(
    modules: Module[],
    moduleDependencies: { moduleName: string; dependencies: string[] }[]
  ) {
    const moduleNodesMap: Record<string, ModuleNode> = {};
    modules.forEach(module => {
      moduleNodesMap[module.name] = { module, children: [], parents: [] };
    });
    moduleDependencies.forEach(moduleDependency => {
      const moduleName = moduleDependency.moduleName;
      moduleDependency.dependencies.forEach(dependency => {
        if (moduleNodesMap[dependency]) {
          moduleNodesMap[moduleName].children.push(moduleNodesMap[dependency]);
          moduleNodesMap[dependency].parents.push(moduleNodesMap[moduleName]);
        }
      });
    });
    return moduleNodesMap;
  }

  private static _sortModules(moduleNodesMap: Record<string, ModuleNode>) {
    const nodes: Node[] = [];

    Object.values(moduleNodesMap).forEach(moduleNode => {
      nodes.push({
        name: moduleNode.module.name,
        children: moduleNode.children.map(c => c.module.name)
      });
    });

    const sortedNodes = bfs(nodes);
    const sortedModuleNames = sortedNodes.map(n => n.name);
    return sortedModuleNames;
  }

  get rootModuleName(): string {
    return this._rootModuleName;
  }

  getModule(moduleName: string): ModuleNode {
    const moduleNode = this._moduleNodesMap[moduleName];
    if (moduleNode === undefined) {
      throw new Error(`module '${moduleName}' not found`);
    }
    return moduleNode;
  }

  get list(): ModuleNode[] {
    return this._moduleNamesInBFSOrder.map(n => this._moduleNodesMap[n]);
  }
}
