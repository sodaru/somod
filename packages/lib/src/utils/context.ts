import { createHash } from "crypto";
import { normalize } from "path";
import {
  IContext,
  IExtensionHandler,
  IModuleHandler,
  INamespaceHandler,
  IServerlessTemplateHandler
} from "somod-types";
import { ExtensionHandler } from "./extension/handler";
import { ModuleHandler } from "./module";
import { NamespaceHandler } from "./namespace";
import { ServerlessTemplateHandler } from "./serverless/serverlessTemplate/serverlessTemplate";

export class Context implements IContext {
  private static instance: IContext;

  private _dir: string;
  private _isUI: boolean;
  private _isServerless: boolean;
  private _isDebugMode: boolean;
  private _moduleHandler: IModuleHandler;
  private _extensionHandler: IExtensionHandler;
  private _namespaceHandler: INamespaceHandler;
  private _serverlessTemplateHandler: IServerlessTemplateHandler;

  private constructor() {
    // do nothing
  }

  static async getInstance(
    dir: string,
    isUI: boolean,
    isServerless: boolean,
    isDebugMode: boolean
  ) {
    if (this.instance === undefined) {
      const context = new Context();
      context._dir = normalize(dir);
      context._isUI = isUI;
      context._isServerless = isServerless;
      context._isDebugMode = isDebugMode;

      context._moduleHandler = await ModuleHandler.getInstance(context._dir);

      context._serverlessTemplateHandler =
        await ServerlessTemplateHandler.getInstance(context);

      context._extensionHandler = await ExtensionHandler.getInstance(
        context._moduleHandler
      );
      context._namespaceHandler = await NamespaceHandler.getInstance(context);

      this.instance = context;
    }
    return this.instance;
  }

  get dir() {
    return this._dir;
  }
  get moduleHandler() {
    return this._moduleHandler;
  }
  get extensionHandler(): IExtensionHandler {
    return this._extensionHandler;
  }
  get namespaceHandler() {
    return this._namespaceHandler;
  }
  get serverlessTemplateHandler() {
    return this._serverlessTemplateHandler;
  }
  get isUI() {
    return this._isUI;
  }
  get isServerless() {
    return this._isServerless;
  }
  get isDebugMode() {
    return this._isDebugMode;
  }
  getModuleHash(moduleName: string) {
    return createHash("sha256")
      .update(moduleName)
      .digest("hex")
      .substring(0, 8);
  }
}
