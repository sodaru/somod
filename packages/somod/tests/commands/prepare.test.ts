import prepareCommand from "../../src/commands/prepare";
import { IContext } from "somod-types";
import * as lib from "somod-lib";
import { mockedFunction } from "../utils";
import { addCommonOptions } from "nodejs-cli-runner";

jest.mock("somod-lib", () => {
  const original = jest.requireActual("somod-lib");
  const mocked = {};
  for (const key in original) {
    mocked[key] =
      typeof original[key] === "function" ? jest.fn() : original[key];
  }
  return {
    __esModule: true,
    ...mocked
  };
});

describe("Test command prepare", () => {
  let output: unknown[][] = [];

  // eslint-disable-next-line no-console
  const originalConsoleLog = console.log;

  const functionsToBeMocked = [
    "bundleFunctionLayers",
    "bundleFunctions",
    "createPages",
    "createPublicAssets",
    "deletePagesAndPublicDir",
    "generateNextConfig",
    "generateRootParameters",
    "initializeContext",
    "prepareSAMTemplate"
  ];

  const mock = (input: Record<string, unknown>) => {
    for (const key of functionsToBeMocked) {
      // eslint-disable-next-line import/namespace
      mockedFunction(lib[key]).mockImplementation(async (...args) => {
        output.push([key, ...args]);
        return input[key];
      });
    }
    mockedFunction(lib.findRootDir).mockImplementation((...args) => {
      output.push(["findRootDir", ...args]);
      return "/root/dir";
    });
  };

  beforeAll(() => {
    addCommonOptions(prepareCommand);
  });

  beforeEach(() => {
    prepareCommand.setOptionValue("verbose", undefined);
    prepareCommand.setOptionValue("debug", undefined);
    prepareCommand.setOptionValue("ui", undefined);
    prepareCommand.setOptionValue("serverless", undefined);

    // eslint-disable-next-line no-console
    console.log = originalConsoleLog;
    for (const key of functionsToBeMocked) {
      // eslint-disable-next-line import/namespace
      mockedFunction(lib[key]).mockReset();
    }
    output = [];

    // eslint-disable-next-line no-console
    console.log = jest.fn();
  });

  test("without args", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          preprepareHooks: [],
          prepareHooks: []
        }
      } as IContext
    });
    await prepareCommand.parseAsync([], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });

  test("with verbose", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          preprepareHooks: [],
          prepareHooks: []
        }
      } as IContext
    });
    await prepareCommand.parseAsync(["-v"], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });

  test("with debug mode", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          preprepareHooks: [],
          prepareHooks: []
        }
      } as IContext
    });
    await prepareCommand.parseAsync(["-d"], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });

  test("with serverless", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          preprepareHooks: [],
          prepareHooks: []
        }
      } as IContext
    });
    await prepareCommand.parseAsync(["--serverless"], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });

  test("with ui", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          preprepareHooks: [],
          prepareHooks: []
        }
      } as IContext
    });
    await prepareCommand.parseAsync(["--ui"], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });

  test("with all options", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          preprepareHooks: [],
          prepareHooks: []
        }
      } as IContext
    });
    await prepareCommand.parseAsync(["-v", "-d", "--ui", "--serverless"], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });

  test("with hooks from extensions", async () => {
    mock({
      initializeContext: {
        dir: "/root/dir",
        extensionHandler: {
          preprepareHooks: [
            {
              extension: "e1",
              value: context => {
                output.push(["pre-e1", context]);
              }
            },
            {
              extension: "e2",
              value: context => {
                output.push(["pre-e2", context]);
              }
            },
            {
              extension: "e3",
              value: context => {
                output.push(["pre-e3", context]);
              }
            }
          ],
          prepareHooks: [
            {
              extension: "e1",
              value: context => {
                output.push(["e1", context]);
              }
            },
            {
              extension: "e2",
              value: context => {
                output.push(["e2", context]);
              }
            },
            {
              extension: "e3",
              value: context => {
                output.push(["e3", context]);
              }
            }
          ]
        }
      } as IContext
    });
    await prepareCommand.parseAsync(["-v", "-d", "--ui", "--serverless"], {
      from: "user"
    });
    expect(output).toMatchSnapshot();
  });
});
