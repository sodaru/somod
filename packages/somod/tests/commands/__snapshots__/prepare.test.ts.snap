// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test command prepare with all options 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    true,
    true,
    true,
  ],
  Array [
    "generateRootParameters",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "deletePagesAndPublicDir",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "createPages",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "createPublicAssets",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "generateNextConfig",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "bundleFunctions",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
    true,
  ],
  Array [
    "bundleFunctionLayers",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
    true,
  ],
  Array [
    "prepareSAMTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
]
`;

exports[`Test command prepare with debug mode 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    true,
    true,
    true,
  ],
  Array [
    "generateRootParameters",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "deletePagesAndPublicDir",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "createPages",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "createPublicAssets",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "generateNextConfig",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "bundleFunctions",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "bundleFunctionLayers",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "prepareSAMTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
]
`;

exports[`Test command prepare with serverless 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    undefined,
    true,
    undefined,
  ],
  Array [
    "generateRootParameters",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "bundleFunctions",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "bundleFunctionLayers",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "prepareSAMTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
]
`;

exports[`Test command prepare with ui 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    true,
    undefined,
    undefined,
  ],
  Array [
    "generateRootParameters",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "deletePagesAndPublicDir",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "createPages",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "createPublicAssets",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "generateNextConfig",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
]
`;

exports[`Test command prepare with verbose 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    true,
    true,
    undefined,
  ],
  Array [
    "generateRootParameters",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "deletePagesAndPublicDir",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "createPages",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "createPublicAssets",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "generateNextConfig",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "bundleFunctions",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
    true,
  ],
  Array [
    "bundleFunctionLayers",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
    true,
  ],
  Array [
    "prepareSAMTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
]
`;

exports[`Test command prepare without args 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    true,
    true,
    undefined,
  ],
  Array [
    "generateRootParameters",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "deletePagesAndPublicDir",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "createPages",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "createPublicAssets",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "generateNextConfig",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
  Array [
    "bundleFunctions",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "bundleFunctionLayers",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
    undefined,
  ],
  Array [
    "prepareSAMTemplate",
    Object {
      "dir": "/root/dir",
      "extensionHandler": Object {
        "prepareHooks": Array [],
        "preprepareHooks": Array [],
      },
    },
  ],
]
`;
